### PR TITLE
[WIP] Implementation of a DynamoDB-based gossip channel

### PR DESCRIPTION
**Pull request by apatenotre**
Hi,

This is an attempt to bring DynamoDB as a gossip channel.
Still a work-in-progress, but I think that some of the main components are now built.

Design is heavily inspired from the already-existing Azure-based gossip channel. I tried to keep the same conventions, but feel free to ask me some changes if this does not comply with your dev rules.

Work is not over yet, as
- not a single test has been done :)
- I have not yet used DynamoDB before, so there may be design errors
- the gossip channel still has to be registered in the MultiClusterGossipChannelFactory

I have made the choice to split data into two different tables: one for the configuration, and one for the gateway. This is not the way it has been implemented in the Azure channel, but I am not at all familiar with AzureTable.
This may or may be not a good choice, feel free to comment

Also, a LegacyDynamoDBMembershipConfigurator has been used to parse DynamoDB connection string. I have to so as IGossipChannel's Initialize makes it mandatory. There may be better ways to do that

### CODE REVIEW

**Comment by apatenotre** [comment]
I've fixed the DynamoDB logic, which was indeed very buggy :)
I am working on the Gossip Channel itself now, checking update & synchronize logic

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/DynamoDBBasedGossipChannel.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+                logger.Debug("-Synchronize pulled delta:{0}", delta);
+
+                return delta;
+            }
+            catch (Exception e)
+            {
+                logger.Info("-Synchronize encountered exception {0}", e);
+
+                throw e;
```
[the remark this answers is not in the document]
Comment (apatenotre) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/DynamoDBBasedGossipChannel.cs]: absolutely - fixed in a commit that is not yet pushed

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        #endregion
+
+        private async Task<bool> TryOperation(Func<Task> func, string operation = null)
+        {
+            try
+            {
+                await func().ConfigureAwait(false);
+                return true;
```
[the remark this answers is not in the document]
Comment (apatenotre) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs]: yes, my feeling is that this should not throw an exception, and instead return false after logging the error, but this is the way it is done currently in the Azure implementation

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipConfiguration.cs [hunk 3]
```diff
@@ -5,6 +5,10 @@
 
 namespace Orleans.Clustering.DynamoDB.MultiClusterNetwork
 {
+    /// <summary>
+    /// Represents a Gossip COnfiguration, as stored in DynamoDB
```
[the remark this answers is not in the document]
Comment (apatenotre) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipConfiguration.cs]: will be fixed in next commit

**Comment by apatenotre** [comment]
Ok now I believe it's in a rather stable state. 
Some questions remain, but I would need feedback from a reviewer to address them

**Comment by apatenotre** [comment]
Hi @sebastianburckhardt , just wondering if you had any chance to have a look at it ?

**Comment by sergeybykov** [comment]
@sebastianburckhardt can you take a look?

**Comment by sebastianburckhardt** [comment]
I have a review deadline today but can look at it tomorrow.

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs]: Yes, this should not throw an exception if the error is caused by any type of concurrency or contention.

For the Azure implementation, note that the exceptions are caught and analyzed, and there is the following line in GossipTableInstanceManager.cs:

```
if (AzureStorageUtils.IsContentionError(httpStatusCode)) return false;
```
which avoids throwing an exception for HttpStatusCode.PreconditionFailed, HttpStatusCode.Conflict, HttpStatusCode.NotFound,HttpStatusCode.NotImplemented.

[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs]: I am not familiar with DynamoDB API, but it probably has its own way of reporting .. a quick look at DynamoDBGrainStorage.cs reveals that there is some exception called `ConditionalCheckFailedException`, which sounds like the thing that should be caught.

**Comment by sebastianburckhardt** [comment]
Overall, looks good to me - I am not familiar with the details of the DynamoDB API but the structure closely follows the one used by Azure table implementation.

I am assuming you did some testing.. but the tests are not currently part of the PR as far as I can tell. It would be great to copy some or all of the tests in `AzureGossipTableTests` (especially the last two), creating a corresponding file in the AWSUtils.Tests project.

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (apatenotre) [reply src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs]: Thanks for pointing this out. I will look into it

**Comment by apatenotre** [comment]
Yes, I did some tests but they were not in an acceptable state, so they have not been commited. 
I will have a look at both AWS and Azure tests currently in the repo and reproduce their structure. Will try to do this within the next days, hopefully by the end of the week

File: src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+        private GossipTableInstanceManager(string globalServiceId, string storageConnectionString, ILoggerFactory loggerFactory)
+        {
+            _globalServiceId = globalServiceId;
+
+            _logger = loggerFactory.CreateLogger<GossipTableInstanceManager>();
+
+            // Using legacy here
+            var options = new DynamoDBClusteringOptions();
+            LegacyDynamoDBMembershipConfigurator.ParseDataConnectionString(storageConnectionString, options);
```
Comment (apatenotre) [inline src/AWS/Orleans.Clustering.DynamoDB/MultiClusterNetwork/GossipTableInstanceManager.cs]: @sebastianburckhardt what is your opinion on this one ? Using Legacy* stuff makes me somehow uncomfortable, but I couldn't find a better way to do it

**Comment by apatenotre** [comment]
Do you mind if I add [Moq](https://www.nuget.org/packages/moq/) to the test project ? It would greatly help for unit tests.
I have had a look at `AzureGossipTableTests` but I'm not sure I will be able to keep the same structure. Besides, I think we can have a better coverage, and Moq will help me for this purpose :)
Also, and still for testing purposes, I have introduced an interface on DynamoDBStorage in the last commit

**Comment by apatenotre** [comment]
> Do you mind if I add Moq to the test project ?

Hi @sebastianburckhardt , @sergeybykov , may I move on in this direction ?
Let me know !

**Comment by apatenotre** [comment]
Hi Sergey, no worries I have been quite busy with some other tasks recently and didn't push a lot for it.
I understand your concern (was expecting it, tbh), and will write some tests without Moq. Should be done within the next days, maybe one or two weeks

**Comment by galvesribeiro** [comment]
Hello, did this PR moved forward?

**Comment by apatenotre** [comment]
Hi @galvesribeiro, code should not have to be changed, but I still have to write some tests. I've been extremely busy and could not find time to do it so far. 
Anyway this PR is something I will need soon, and that's a good incentive to finish the job :)
Let me try to move on within the next two weeks

**Comment by galvesribeiro** [comment]
Sure! Take your time!

**Comment by galvesribeiro** [comment]
@apatenotre is it done or do you need more time before a final review? I'm out on vacation but will take a look on 23rd when I get back unless someone from the team look at it and merge first.

**Comment by apatenotre** [comment]
Not yet, I added the WIP flag back. Code is fine, just need to write some
tests. That will be done shortly, le me ping you when it's over. Thanks!

Le lun. 13 mai 2019 à 23:23, Gutemberg Ribeiro <notifications@github.com> a
écrit :

**Comment by sergeybykov** [comment]
@apatenotre Are you still pursuing this PR?

**Comment by sergeybykov** [comment]
Closing for now due to inactivity. Feel free to reopen.

**Pull request closed without merging**